### PR TITLE
Update dropdown conditions on column rename

### PR DESCRIPTION
**Pull request by SleepyLeslie**
This PR adds the ability to automatically update the dropdown condition formula on a reference column when a column it refers to has been renamed.

### CODE REVIEW

**Comment by SleepyLeslie** [comment]
Refactored the entire logic. Now ACL formulas and dropdown condition formulas use the same logic to handle renames. Previously `$` notations in ACL formulas are not auto-renamed. This should also be fixed now.

Tests to come later.

File: sandbox/grist/predicate_formula.py
Comment (jordigh) [inline sandbox/grist/predicate_formula.py:116]: This works, but it feels a little too manual and verbose. Wouldn't you rather use the `re` module?
```python
rec_occurrences = (m.start() for m in re.finditer("rec.", new_formula_text))
```
(Note the use of round brackets `()` and not square `[]` so that we can benefit from lazy evaluation in line 119.)

**Review by jordigh — COMMENTED** [review]
I only have superficial style comments right now. I'm still trying to understand the deeper structure of this change.

File: sandbox/grist/predicate_formula.py
Comment (SleepyLeslie) [reply sandbox/grist/predicate_formula.py:50]: I always catch errors under the name `e`, and for consistency shortened it on the spot when I saw it. Minor thing though?

Good point for the f-string.

File: sandbox/grist/dropdown_condition.py
Comment (SleepyLeslie) [reply sandbox/grist/dropdown_condition.py:42]: I was trying to catch failed `json.loads` due to `col.widgetOptions` being empty, and failed JSON property access at any level. Wasn't sure what else could go wrong, but in general I considered any error an indicator of a column without proper `widgetOptions`.

Should be `json.JSONDecodeError` and `KeyError` if it were to be explicit. Anything I might have missed?

File: sandbox/grist/acl.py
Comment (SleepyLeslie) [reply sandbox/grist/acl.py:51]: Oh, I just thought it'd be redundant to pass `docmodel` alongside `useractions` and didn't realize there's a convention about the underscore prefix. I want a second opinion on what to do about this.

File: sandbox/grist/acl.py
Comment (SleepyLeslie) [reply sandbox/grist/acl.py:170]: They are different. `ValueError` is raised when there is valid but _unsupported_ syntax. It's raised by `TreeConverter`. `SyntaxError` is raised by `ast.parse` when there is invalid syntax.

I was also confused by this.

File: sandbox/grist/predicate_formula.py
Comment (SleepyLeslie) [reply sandbox/grist/predicate_formula.py:116]: Beautiful code. One small gotcha: pattern should be `r"rec\."`.

File: sandbox/grist/predicate_formula.py
Comment (paulfitz) [reply sandbox/grist/predicate_formula.py:51]: (F-string warning: the data engine still needs to be Python 2.7 compatible because of some legacy users, sorry! There can be exceptions, with some care, but right here seems a bad place for that exception.)

Comment (SleepyLeslie) [reply sandbox/grist/predicate_formula.py:51]: Reverted the f-string change.

File: sandbox/grist/dropdown_condition.py
Comment (georgegevoian) [inline sandbox/grist/dropdown_condition.py:46]: Something to watch out for here: Reference List columns have a different prefix (`RefList:`).

File: sandbox/grist/dropdown_condition.py
Comment (georgegevoian) [inline sandbox/grist/dropdown_condition.py:34]:
```suggestion
  for col in useractions._docmodel.columns.all:
```

File: sandbox/grist/dropdown_condition.py
Comment (georgegevoian) [inline sandbox/grist/dropdown_condition.py:38]:
```suggestion
    # Find all columns in the document that have dropdown conditions.
```

File: sandbox/grist/dropdown_condition.py
Comment (georgegevoian) [inline sandbox/grist/dropdown_condition.py:29]:
```suggestion
  Given a dict of column renames of the form {(table_id, col_id): new_col_id}, applies updates
```

File: sandbox/grist/test_dropdown_condition_renames.py
Comment (georgegevoian) [inline sandbox/grist/test_dropdown_condition_renames.py:46]: Thanks for adding a test case.

File: sandbox/grist/dropdown_condition.py
Comment (georgegevoian) [inline sandbox/grist/dropdown_condition.py:51]: Choice and Choice List columns support dropdown conditions too. They work a little different: `choice` is just a string (and doesn't have any attributes that need renaming). But we should still rename any instances of `$`/`rec` that need renaming.

Actually, it would be good to always rename `$`/`rec`, regardless of column type. Dropdown conditions stay in widget options, even when the current column type can't make use of them. You can see this when you cycle between column types that support and don't support dropdown conditions. For columns that aren't Choice or Reference [Lists], it's not possible to know whether `choice` was a string or a record, but we can still rename the `$`/`rec` instances, if needed.

Comment (SleepyLeslie) [reply sandbox/grist/dropdown_condition.py]: Thanks for the info! I made some changes and added a test case for the situation you mentioned. Hopefully that addresses your comment.

File: sandbox/grist/acl.py
Comment (SleepyLeslie) [reply sandbox/grist/acl.py:51]: I feel it's unnecessary to make a new class for this purpose. I'll just remove the underscore from `useractions._docmodel`.

**Review by dsagal — COMMENTED** [review]
Also, could you rebase to (or merge from) the latest `main` branch please? There have been a couple of changes to Python code, which shouldn't cause conflicts, but it would be good to check before landing.

File: sandbox/grist/acl.py
Comment (SleepyLeslie) [reply sandbox/grist/acl.py:172]: I ran pylint manually with C0303 and W293 and fixed all reported occurrences. Hope it's enough for now.

File: sandbox/grist/acl.py
Comment (SleepyLeslie) [reply sandbox/grist/acl.py:6]: No. I checked for and fixed all unused imports with pylint.

File: sandbox/grist/dropdown_condition.py
Comment (SleepyLeslie) [reply sandbox/grist/dropdown_condition.py:62]: Makes sense!

Comment (SleepyLeslie) [reply sandbox/grist/dropdown_condition.py:62]: On second thought, if more fields get added in the future, we'll need to be careful about possible inconsistencies if the new fields are not updated accordingly here.

I don't have a preference here though. Will take your suggestion.

File: sandbox/grist/acl.py
Comment (SleepyLeslie) [reply sandbox/grist/acl.py:165]: I don't believe this can result in inconsistency. Even if `aclFormula` is updated, it will still be broken if it originally was. The parsed version will stay obsolete and broken until the text version is fixed by the user, when the parsed version is re-generated and fixed thereof.

But I went through the logic again and recalled that the new `process_renames` function doesn't do anything to a formula with syntax errors, so nothing will actually happen. Error checking here seems indeed redundant. I removed them and left comments to explain the situation.

Also: I have a test case for this behavior (don't process syntactically wrong formulas), so we should be safe.

File: sandbox/grist/dropdown_condition.py
Comment (SleepyLeslie) [reply sandbox/grist/dropdown_condition.py:66]: I removed the redundant error checking here as well.

File: sandbox/grist/predicate_formula.py
Comment (SleepyLeslie) [reply sandbox/grist/predicate_formula.py:93]: I tried doing that initially. I remember giving up because the `ValueError` was raised by `TreeConverter`, which I thought might be used elsewhere.

A revisit of this issue suggests it might actually be good to raise `SyntaxError` instead, as most code using `TreeConverter` has been merged, and `predicate_formula.py` has become its sole user.

File: sandbox/grist/dropdown_condition.py
Comment (SleepyLeslie) [reply sandbox/grist/dropdown_condition.py]: I'll mark this one as resolved and address the issue together with the dropdown condition, since they are essentially the same.

File: sandbox/grist/dropdown_condition.py
Comment (SleepyLeslie) [reply sandbox/grist/dropdown_condition.py]: Will do!

Comment (SleepyLeslie) [reply sandbox/grist/dropdown_condition.py]: `useractions.py` uses `col.type.split(':')[:2]` heavily. I'm not sure what the `[:2]` is for (what can column types be that contains two or more colons?) Also this pattern doesn't match my use case where I am not sure whether the type looks like `Ref[List]:x`. I'll leave the old code alone.

File: sandbox/grist/acl.py
Comment (SleepyLeslie) [reply sandbox/grist/acl.py:51]: @dsagal Which do you prefer? Reverting, making two new classes, or adding an accessor?

File: sandbox/grist/predicate_formula.py
Comment (SleepyLeslie) [reply sandbox/grist/predicate_formula.py:97]: Thank you! I jumped too quickly to `map_back_offset`, should've looked more closely. This is so nice.

File: sandbox/grist/dropdown_condition.py
Comment (dsagal) [inline sandbox/grist/dropdown_condition.py:57]: This sentence sounds wrong. Generally Grist formulas can have exceptions and that doesn't normally interfere with other formulas. So probably you mean something more specific -- it would be good to clarify.

File: sandbox/grist/predicate_formula.py
Comment (dsagal) [inline sandbox/grist/predicate_formula.py:52]: Cool! I didn't know about `raise...from`!

File: sandbox/grist/test_dropdown_condition_renames.py
Comment (dsagal) [inline sandbox/grist/test_dropdown_condition_renames.py:114]: Is this actually still needed and intentional? Doesn't seem right to swallow errors silently.

**Review by dsagal — COMMENTED** [review]
Thank you! Looks good, just one lingering question about syntaxerrors.

File: sandbox/grist/predicate_formula.py
Comment (berhalak) [reply sandbox/grist/predicate_formula.py:52]: But I think it doesn't work in python 2, so we can't use it.

File: sandbox/grist/test_dropdown_condition_renames.py
Comment (SleepyLeslie) [reply sandbox/grist/test_dropdown_condition_renames.py:114]: No longer needed. It used to pair with another solution I didn't settle on.

File: sandbox/grist/predicate_formula.py
Comment (SleepyLeslie) [reply sandbox/grist/predicate_formula.py:52]: Oops, that's sad... Using `six.reraise` instead.

File: sandbox/grist/dropdown_condition.py
Comment (SleepyLeslie) [reply sandbox/grist/dropdown_condition.py]: I rephrased it a bit. Hope it's clearer now.

File: sandbox/grist/dropdown_condition.py
Comment (dsagal) [reply sandbox/grist/dropdown_condition.py]: I thought there might be something with DateTime that may contains two spaces, but now that I think about it, I don't think so. Maybe it just avoids assuming more. `col.type.split(':')[1]` for Ref or RefList columns should work, but you approach works too, so up to you.

**Comment by dsagal** [comment]
I think one remaining comment is about making public a `UserActions` member that seems better to keep non-public: https://github.com/gristlabs/grist-core/pull/1038#discussion_r1667994619

File: sandbox/grist/acl.py
Comment (dsagal) [reply sandbox/grist/acl.py:51]: Sorry, I didn't see your question in my last comment. No strong preference, so I suggest an accessor as a simpler change closer to what you already have.

**Review by dsagal — APPROVED** [review]
I have two more comments but really minor now. Approving. Thank you very much! I appreciate in particular unifying quite a bit of the logic and fixing the renaming of acl formulas with "$" signs along the way!

File: sandbox/grist/acl.py
Comment (SleepyLeslie) [reply sandbox/grist/acl.py:152]: That's fair. I have no strong preference and will do what you suggested.

File: sandbox/grist/useractions.py
Comment (berhalak) [inline sandbox/grist/useractions.py:711]: @dsagal @SleepyLeslie Can I move those two methods down below (where it was before)? 
The new order of things generates acl rename actions before the column is actually renamed. Before the order was reversed, first `RenameAction` then `UpdateRecord` (for ACL)

Comment (berhalak) [reply sandbox/grist/useractions.py:711]: After discussing it offline, I will update it in another PR.

**Pull request merged**
